### PR TITLE
HTML API: Use full parser for html5lib tests

### PR DESCRIPTION
**Pull request by sirreal**
Trac ticket: Core-61646

Previously, the html5lib tests have only run in the fragment parser mode, but in #6977 a full parser was introduced to the HTML API. In this patch, the html5lib tests are running in the full parser mode where applicable.

#7146 is merged here and should be landed first.

```diff
- Tests: 609, Assertions: 405, Skipped: 204.
+ Tests: 1498, Assertions: 787, Skipped: 711.
```

This vastly increases the tests that are used from HTML5lib-tests from 609 to 1498! It's helped to discover two issues:

- https://core.trac.wordpress.org/ticket/61831
- https://core.trac.wordpress.org/ticket/61810

### CODE REVIEW

File: tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php [hunk 1]
```diff
@@ -68,14 +74,43 @@ class Tests_HtmlApi_Html5lib extends WP_UnitTestCase {
...
+			}
+		} elseif ( str_ends_with( $expected_tree, $auto_generated_head_body ) && ! str_ends_with( $processed_tree, $auto_generated_head_body ) ) {
+			if ( str_ends_with( $processed_tree, "<head>\n\n" ) ) {
+				$processed_tree = substr_replace( $processed_tree, "  <body>\n\n", -1 );
+			} else {
+				$processed_tree = substr_replace( $processed_tree, $auto_generated_head_body, -1 );
+			}
+		} elseif ( str_ends_with( $expected_tree, $auto_generated_body ) && ! str_ends_with( $processed_tree, $auto_generated_body ) ) {
+			$processed_tree = substr_replace( $processed_tree, $auto_generated_body, -1 );
+		}
```
Comment (sirreal) [inline tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php:89]: `str_ends_with` will need to be replaced with something PHP-7.2 compatible. It's a PHP8 feature.

Comment (dmsnell) [reply tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php:89]: we polyfill it in WordPress - it should be find here, unless we don't load Wordpress

File: src/wp-includes/html-api/class-wp-html-tag-processor.php [hunk 2]
```diff
@@ -3532,7 +3545,13 @@ private function matches(): bool {
...
 
 		// Does the tag name match the requested tag name in a case-insensitive manner?
-		if ( isset( $this->sought_tag_name ) && 0 !== substr_compare( $this->html, $this->sought_tag_name, $this->tag_name_starts_at, $this->tag_name_length, true ) ) {
+		if (
+			isset( $this->sought_tag_name ) &&
+			(
+				strlen( $this->sought_tag_name ) !== $this->tag_name_length ||
+				0 !== substr_compare( $this->html, $this->sought_tag_name, $this->tag_name_starts_at, $this->tag_name_length, true )
+			)
+		) {
```
Comment (dmsnell) [inline src/wp-includes/html-api/class-wp-html-tag-processor.php]: we need this. `next_tag( 'PLAINTEXT' )` stopped on `P`

**Comment by dmsnell** [comment]
@sirreal I've added active format reconstruction to the `PLAINTEXT` element, even though it's missing from the spec, but all the major parsers clearly perform this. I've submitted a PR to the HTML spec to add the rule.

https://github.com/whatwg/html/pull/10540

**Comment by dmsnell** [comment]
Leaving a note for posterity's sake: adding active format reconstruction to PLAINTEXT is wrong. Adding PLAINTEXT is more complicated than it originally seemed, because even though every remaining token is a character token, if there are any character tokens left, they may trigger active format reconstruction. This is something I'll try and fix

**Comment by dmsnell** [comment]
Fragment support extracted into https://github.com/WordPress/wordpress-develop/pull/7141 where I thought I had already done so, sorry!

**Comment by sirreal** [comment]
Thanks, I'll revert the changes to support arbitrary fragments here and continue to only support `body`. That work is more complicated and can continue in #7141.

File: tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php [hunk 3]
```diff
@@ -68,14 +55,43 @@ class Tests_HtmlApi_Html5lib extends WP_UnitTestCase {
...
 		}
+		$fragment_detail = $fragment_context ? " in context <{$fragment_context}>" : '';
 
-		$this->assertSame( $expected_tree, $processed_tree, "HTML was not processed correctly:\n{$html}" );
+		/*
+		 * The HTML processor does not produce html, head, body tags if the processor does not reach them.
+		 * These should all be produced when reaching the end-of-file.
+		 * For now, append the missing tags when necessary.
+		 *
+		 * @todo remove this section when when the processor handles this.
```
Comment (sirreal) [inline tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php]:
```suggestion
		 * HTML tree construction will always produce these tags, the HTML API does not at this time.
```

**Review by dmsnell — APPROVED** [review]
This is a nice improvement in how many tests run, and it's not trivial to fix the test runner. Thanks again @sirreal!

This is a tests-only change.

**Comment by dmsnell** [comment]
Merged in [[58859]](https://core.trac.wordpress.org/changeset/58859)
https://github.com/wordpress/wordpress-develop/commit/9928cd6bcb35d14a700e3930355b6703f0cb4dc8

**Pull request closed without merging**